File: genja/examples/run_task.rs
use genja::genja_core::inventory::Host;
use genja::genja_core::task::{
    HostTaskResult, Task, TaskError, TaskRuntimeContext, TaskSuccess,
};
use genja::{Genja, genja_task};
use serde_json::json;

struct CollectFacts;

#[genja_task(name = "collect_facts")]
impl CollectFacts {
    async fn start_async(
        &self,
        host: &Host,
        _context: &TaskRuntimeContext,
    ) -> Result<HostTaskResult, TaskError> {
        Ok(HostTaskResult::passed(TaskSuccess::new().with_result(
            json!({
                "hostname": host.hostname(),
                "platform": host.platform(),
                "facts_collected": true
            }),
        )))
    }
}

fn main() -> Result<(), genja::GenjaError> {
    let genja = Genja::from_settings_file("genja/examples/settings.yaml")?;

    let results = genja.run_task(
        CollectFacts,
        1,
    )?;

    let output = results.to_pretty_json_string().map_err(|err| {
        genja::GenjaError::Message(format!("failed to serialize task results: {err}"))
    })?;
    println!("{output}");

    Ok(())
}

File: genja/examples/python/run_task.py
from pathlib import Path
import json

import genja
from genja.task import TaskSuccessResult, task


EXAMPLES_DIR = Path(__file__).resolve().parents[1]
HOSTS_FILE = EXAMPLES_DIR / "inventory" / "hosts.json"


@task(name="collect_facts")
class CollectFacts:
    def run(self, task, host, context):
        return TaskSuccessResult(
            summary=f"collected facts from {host.hostname}",
            metadata={
                "hostname": host.hostname,
                "platform": host.platform,
                "facts_collected": True,
            },
        )


with HOSTS_FILE.open() as hosts_file:
    hosts = json.load(hosts_file)

runtime = genja.Genja.from_hosts(hosts).with_runner("serial")
results = runtime.run_task(CollectFacts, max_depth=1)

print(results.to_json(pretty=True))

